text/0000-disconnect-flow.md

```diff
@@ -0,0 +1,134 @@
+  
+  ```C#
+  void OnServerConnectionLost(NetworkManager.DisconnectReason reason)
+  ```
+
+The `reason` parameter is an enum currently defined as follows:
+
+```C#
+public enum DisconnectReason
+{
+    TimeOut, // Connection to the remote process timed out
+    Stopped, // The remote process shut down its network cleanly
+    Kicked, // The client was kicked from the server
+    ConfigurationMismatch, // The client and server have mismatched configurations
+    NotAuthorized, // The client is not (or is no longer) authorized to be connected
+}
+```
+
+# Reference-level explanation
+
+There are basically three components to this implementation.
+
+The first component is to enable to server to broadcast to clients which other clients are connected. This is implemented with two commands: a new `CLIENT_CONNECTED` command and the existing `CONNECTION_APPROVED` command. When a new client connects, a `CLIENT_CONNECTED` message with that clients ID is broadcast to all other connected clients to inform them of the new peer, and when the `CONNECTION_APPROVED` command is sent to the new peer, a list of existing connected client IDs is included. 
+
+ Upon receipt of one of these messages:
+
+- If the message is `CLIENT_CONNECTED`, the `OnPeerConnectedCallback` is invoked.
+
+- If the message is `CONNECTION_APPROVED`, the `OnServerConnectionEstablished` callback is invoked. **This is a change: The client will no longer receive the OnClientConnected callback, which is now server-only.**
+
+The other new message that's added is a `DISCONNECT` message, which contains in it the `DisconnectReason` value and the disconnecting client's clientId. A `DISCONNECT` message is sent:
+
+- From client to server any time `NetworkManager.StopClient()` or `NetworkManager.StopHost()` is called, providing `DisconnectReason.ShutDown` as the reason. **Server does not honor or use the clientId value for security reasons. It infers this value from the identity of the sender.**
+
+- From server to all clients with the `ServerClientId` any time `NetworkManager.StopServer()` or `NetworkManager.StopHost()` is called, providing `DisconnectReason.ShutDown` as the reason
+
+- From server to all clients with the disconnecting client's clientId any time it receives a `DISCONNECT` from a client, forwarding the received reason
+
+- From server to all clients with the disconnecting client's clientId any time connection to a client times out, providing `DisconnectReason.TimedOut` as the reason
+
+- From server to a client that has sent a `CONNECTION_REQUEST` message with a mismatched configuration, providing `DisconnectReason.ConfigurationMismatch` as the reason
+
+- From server to a client that has sent a `CONNECTION_REQUEST` message that's rejected by the `ConnectionApprovalCallback`, providing `DisconnectReason.NotAuthorized` as the reason
```

Comment:
Connection approval isn't necessarily about authorization though. You could have a game check to make sure you don't have too many players of a certain type or coming from a certain region (for whatever reason). That wouldn't be "authorization"

Reply:
True, in line with the comment about HTTP-based error codes above, we can change the connection approval handler to return a reason for not approving. Possibly also integrate the user data as well with that.